## Programming part of Homework 2 (Data Structures, Spring 2021)

## Name:蔡翔宇
## Student ID Number:108820018

### Programming problem 1
Recall that in problem 1, give a recursive algorithm to compute the product of two positive integers, $n$ and $m$, using only addition and subtraction.
This programming problem is to consider the recursive approach in above problem and asks you to complete the following tasks.
1. Implement the approach as a function named as `product_rec` using Python.
2. Please have an iterative version for the approach and write a function for this version with function name `product_ite`.
3. Compare these two function with the same input in terms of running time and write what you have observed.

Note that we will use an in-built python library **`timeit`** and the module function **`timeit.timeit()`** for measuring the running time.

In [22]:
# Please write your code for product_rec() here
def product_rec(a, b):
    if a < b:
        return product_rec(b, a) 
    elif b == 0:
        return 0
    else:
        return a + product_rec(a, b-1)

# print(product_rec(100, 100))
    


10000


In [21]:
# Please write your code for product_ite()
def product_ite(a, b):
    product = 0
    if a < b :
        return product_ite(b, a)
    for i in range(b):
        product += a
    return product

# print(product_ite(100, 100))

10000


In [22]:
# This piece of code is for the ways to measure the running time.
# Please check the document of timeit library for more detail.
# We here briefly give two methods for your reference.
#---------------------------------------------------------timeit
import timeit as ti

#--------------------------- function for illustration
def product_rec(a, b):
    if a < b:
        return product_rec(b, a) 
    elif b == 0:
        return 0
    else:
        return a + product_rec(a, b-1)

def product_ite(a, b):
    product = 0
    if a < b :
        return product_ite(b, a)
    for i in range(b):
        product += a
    return product

#---------------------------input
x = int(input("Enter X:"))
y = int(input("Enter Y:"))
r = int(input("Enter runs:"))

#---------------------------------------------------------Method 1
print("Running time comparison (Timer Method 1):")
tstart = ti.default_timer() #timer start 計時器開啟
for i in range(r):
    product_rec(x, y)
tend = ti.default_timer() #timer end計時器關閉
t_rec=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(r):
    product_ite(x, y)
tend = ti.default_timer() #timer end計時器關閉
t_ite=tend - tstart#開始與結束的差值
print("Recursion result", product_rec(x, y),"Iteration result", product_ite(x, y))
print("Recursive approach:", t_rec)
print("iterative approach:", t_ite)

#---------------------------------------------------------Method 2
print("Running time comparison (Timer Method 2):")
t_rec = ti.timeit("product_rec(259, 767)", setup="from __main__ import product_rec", number = r)#timeit(函數名稱,執行次數)
t_ite = ti.timeit("product_ite(259, 767)", setup="from __main__ import product_ite", number = r)

print("Recursion result", product_rec(x, y),"Iteration result", product_ite(x, y))
print("Recursive approach:", t_rec)
print("Iterative approach:", t_ite)

Running time comparison (Timer Method 1):
Recursion result 198653 Iteration result 198653
Recursive approach: 0.7089250999997603
iterative approach: 0.1744077000003017
Running time comparison (Timer Method 2):
Recursion result 198653 Iteration result 198653
Recursive approach: 0.617764000000534
Iterative approach: 0.1548376000000644


#### Please write down the observations you have after the comparisons:
##### Inputs for five experiments
1.(518, 767), 10000 times

2.(1034, 4966), 10000 times

3.(540, 644), 10000 times

4.(540, 644), 500 times

5.(259, 767), 10000 times
##### Results of comparisoms
*******************************************************************************************************
**Enviorment: CPU: Intel Core i7-6700HQ @2.60GHz, RAM: 16GB DDR4 2666MHz, OS: Windows 10 64bit home**

**Method 1**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Recursion      | 1.355486100000 | 2.839510799999 | 1.305371499999 | 0.067739999999 | 0.708925099999 |
|Iteration      | 0.370131600000 | 0.666280299999 | 0.343678699999 | 0.020159499999 | 0.174407700000 |

**Method 2**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Recursion      | 1.361380399999 | 2.884137300000 | 1.261235699999 | 0.067092300000 | 0.617764000000 |
|Iteration      | 0.327444700000 | 0.721376400000 | 0.374187100000 | 0.015829799999 | 0.154837600000 |

*******************************************************************************************************

##### Observations
I found out that the execution time doubles when the numbers of recursion doubled, but the execution time for the iteration still remained very low.

### Programming problem 2
Consider the ***selection sort*** we discussed in class. Please complete the following tasks.
1. Implement the iterative approach using Python and name the function as `Iterative_Selection_Sort`.
2. Please implement the recursive version you provide in problem 2(b) with the function name of `Recursive_Selection_Sort`.
3. Compare these two function with the same input in terms of running time and write what you have observed.

In [42]:
# Please write your code for Iterative_Selection_Sort(A)
def Iterative_Selection_Sort(A):
    A_copy = []
    for i in range(len(A)):
        A_copy.append(A[i])
    
    i = 0   #start index
    temp = 0
    for i in range(len(A_copy)):
        j = i       #min value index
        for k in range(i+1, len(A_copy)):
            if A_copy[k] < A_copy[j]:
                j = k
                # print(j)
        temp = A_copy[i]
        A_copy[i] = A_copy[j]
        A_copy[j] = temp
        # print("No." + str(i+1) + " " + str(A))
    return A


In [26]:
# Please write your code for Recursive_Selection_Sort(A):
def Recursive_Selection_Sort(A):
    A_copy = []
    for i in range(len(A)):
        A_copy.append(A[i])

    def find_min_index(A, index, min_index):
        if index == len(A):
            # print("Min index:" + str(min_index))
            return min_index
        
        elif A[index] <= A[min_index]:
            min_index = index
            return find_min_index(A, index+1, min_index)
        
        elif A[min_index] < A[index]:
            return find_min_index(A, index+1, min_index)

    def sort(A, pos):
        position = pos
        if pos == len(A):
            return A
        else:
            min = find_min_index(A, position+1, position)
            temp = A[min]
            A[min] = A[position]
            A[position] = temp
            return sort(A, pos+1)
    
    return sort(A_copy, 0)


[5, 9, 2, 8, 1, 1, 6, 4, 7, 9, 3, 10]
[1, 1, 2, 3, 4, 5, 6, 7, 8, 9, 9, 10]
[5, 9, 2, 8, 1, 1, 6, 4, 7, 9, 3, 10]


In [44]:
#---------------------------------------------------------timeit
import timeit as ti

#--------------------------------input list

A = [35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82, 249, 421, 118, 203, 405, 26, 69, 400, 358, 50, 21, 33]
B = [249, 421, 118, 203, 405, 26, 69, 400, 358, 50, 21, 33]
C = [453, 212, 84, 340, 310, 95, 161, 262, 291, 94, 270, 106, 106, 343, 406, 63, 144, 286, 197, 250, 66, 59, 20, 186, 303, 184, 88, 278, 310, 447, 16, 239, 170, 364, 366, 316]
D = [106, 343, 406, 63, 144, 286]
E = [303, 184, 88, 278, 310, 447, 16, 239, 170, 364, 366, 316, 63, 144, 286]

#--------------------------------function

def Iterative_Selection_Sort(A):
    A_copy = []
    for i in range(len(A)):
        A_copy.append(A[i])
    
    i = 0   #start index
    temp = 0
    for i in range(len(A_copy)):
        j = i       #min value index
        for k in range(i+1, len(A_copy)):
            if A_copy[k] < A_copy[j]:
                j = k
                # print(j)
        temp = A_copy[i]
        A_copy[i] = A_copy[j]
        A_copy[j] = temp
        # print("No." + str(i+1) + " " + str(A))
    return A_copy

def Recursive_Selection_Sort(A):
    A_copy = []
    for i in range(len(A)):
        A_copy.append(A[i])

    def find_min_index(A, index, min_index):
        if index == len(A):
            # print("Min index:" + str(min_index))
            return min_index
        
        elif A[index] <= A[min_index]:
            min_index = index
            return find_min_index(A, index+1, min_index)
        
        elif A[min_index] < A[index]:
            return find_min_index(A, index+1, min_index)

    def sort(A, pos):
        position = pos
        if pos == len(A):
            return A
        else:
            min = find_min_index(A, position+1, position)
            temp = A[min]
            A[min] = A[position]
            A[position] = temp
            return sort(A, pos+1)
    
    return(sort(A_copy, 0))

#---------------------------------------------------------Method 1
print("Running time comparison (Timer Method 1):")
tstart = ti.default_timer()         #timer start 計時器開啟

for i in range(10000):
    Recursive_Selection_Sort(E)     #run function for 10000 times

tend = ti.default_timer()           #timer end計時器關閉
t_rec = tend - tstart                 #開始與結束的差值

tstart = ti.default_timer()         #timer start 計時器開啟
for i in range(10000):
    Iterative_Selection_Sort(E)

tend = ti.default_timer()           #timer end計時器關閉
t_ite = tend - tstart                 #開始與結束的差值
print(" The sorted array of", E)
print("     with Recursion is", Recursive_Selection_Sort(E))
print(" The sorted array of", E)
print("     with Iteration is", Iterative_Selection_Sort(E))

print("\nRecursive approach: ", t_rec)
print("Iterative approach: ", t_ite)

#---------------------------------------------------------Method 2

print("\nRunning time comparison (Timer Method 2):\n")

t_rec = ti.timeit("Recursive_Selection_Sort(E)", setup="from __main__ import Recursive_Selection_Sort, E", number = 10000)#timeit(函數名稱,執行次數)
t_ite = ti.timeit("Iterative_Selection_Sort(E)", setup="from __main__ import Iterative_Selection_Sort, E", number = 10000)

print(" The sorted array of", E)
print("     with Recursion is", Recursive_Selection_Sort(E))
print(" The sorted array of", E)
print("     with Iteration is", Iterative_Selection_Sort(E))

print("\nRecursive approach: ", t_rec)
print("Iterative approach: ", t_ite)

Running time comparison (Timer Method 1):
 The sorted array of [303, 184, 88, 278, 310, 447, 16, 239, 170, 364, 366, 316, 63, 144, 286]
     with Recursion is [16, 63, 88, 144, 170, 184, 239, 278, 286, 303, 310, 316, 364, 366, 447]
 The sorted array of [303, 184, 88, 278, 310, 447, 16, 239, 170, 364, 366, 316, 63, 144, 286]
     with Iteration is [16, 63, 88, 144, 170, 184, 239, 278, 286, 303, 310, 316, 364, 366, 447]

Recursive approach:  0.52656879999995
Iterative approach:  0.19337129999985336

Running time comparison (Timer Method 2):

 The sorted array of [303, 184, 88, 278, 310, 447, 16, 239, 170, 364, 366, 316, 63, 144, 286]
     with Recursion is [16, 63, 88, 144, 170, 184, 239, 278, 286, 303, 310, 316, 364, 366, 447]
 The sorted array of [303, 184, 88, 278, 310, 447, 16, 239, 170, 364, 366, 316, 63, 144, 286]
     with Iteration is [16, 63, 88, 144, 170, 184, 239, 278, 286, 303, 310, 316, 364, 366, 447]

Recursive approach:  0.5401037999999971
Iterative approach:  0.1788022000

#### Please write down the observations you have after the comparisons:
##### Inputs for five experiments
1.[35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82, 249, 421, 118, 203, 405, 26, 69, 400, 358, 50, 21, 33]    (24 elements)

2.[249, 421, 118, 203, 405, 26, 69, 400, 358, 50, 21, 33]    (12 elements)

3.[453, 212, 84, 340, 310, 95, 161, 262, 291, 94, 270, 106, 106, 343, 406, 63, 144, 286, 197, 250, 66, 59, 20, 186, 303, 184, 88, 278, 310, 447, 16, 239, 170, 364, 366, 316]    (36 elements)

4.[106, 343, 406, 63, 144, 286]

5.[303, 184, 88, 278, 310, 447, 16, 239, 170, 364, 366, 316, 63, 144, 286]
##### Results of comparisoms
*******************************************************************************************************
**Method 1**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Recursion      | 1.204780500000 | 0.367419800000 | 2.788422199999 | 0.131518700000 | 0.526568799999 |
|Iteration      | 0.385021999998 | 0.163065199998 | 0.837904099998 | 0.081658199998 | 0.193371299999 |

**Method 2**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Recursion      | 1.041743000001 | 0.292713500000 | 2.520810300000 | 0.127677399999 | 0.540103799999 |
|Iteration      | 0.359692099998 | 0.137118900000 | 0.754729500000 | 0.062918300000 | 0.178802200000 |

*******************************************************************************************************

##### Observations
When the unsorted list length doubled, the execution time of the iteration method is around double of the original one.
But when it comes to the recursion, the execution time is squared.